include/watchdog.h

```diff
@@ -1,116 +1,206 @@
 /*
- * Copyright (c) 2015 Intel Corporation.
+ * Copyright (c) 2017 Nordic Semiconductor ASA
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/**
+ * @file
+ * @brief Public API for watchdog drivers.
+ */
+
+#if defined(CONFIG_WDT_LEGACY_API)
+
+/*
+ * This is the default, and will be the way until the new API below
+ * will be enforced everywhere.
+ */
+#include <watchdog_legacy.h>
+
+#else
+
 #ifndef _WDT_H_
 #define _WDT_H_
 
 #include <zephyr/types.h>
-#include <device.h>
 #include <misc/util.h>
+#include <device.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-#define WDT_MODE		(BIT(1))
-#define WDT_MODE_OFFSET         (1)
-#define WDT_TIMEOUT_MASK        (0xF)
+/**
+ * @brief Enable calling application callback function on watchdog timeout.
+ */
+#define WDT_MODE_CALLBACK	BIT(0)
 
-enum wdt_mode {
-	WDT_MODE_RESET = 0,
-	WDT_MODE_INTERRUPT_RESET
+/**
+ * @brief Pause watchdog timer when CPU is in sleep state.
+ */
+#define WDT_PAUSE_IN_SLEEP	BIT(1)
+
+/**
+ * @brief Pause watchdog timer when CPU is halted by the debugger.
+ */
+#define WDT_PAUSE_HALTED_BY_DBG		BIT(2)
+
+/**
+ * @brief Watchdog timeout window.
+ *
+ * Each installed timeout needs reloading in specified time window, otherwise
+ * the watchdog will trigger.
+ *
+ * @param up Upper limit of watchdog reload timeout in miliseconds.
+ * @param down Lower limit of watchdog reload timeout in miliseconds.
+ */
+struct wdt_window {
+	u16_t up;
+	u16_t down;
 };
 
 /**
- * WDT clock cycles for timeout type.
+ * @brief Watchdog configuration struct.
+ *
+ * @param options - a bitfield with following parts:
+ *    enable callback     [ 0 ]       - reset or callback -> reset
+ *    pause in sleep      [ 1 ]       - pause in CPU sleep
+ *    pause by debugger   [ 2 ]       - pause when halted by debugger
  */
-enum wdt_clock_timeout_cycles {
-	WDT_2_16_CYCLES,
-	WDT_2_17_CYCLES,
-	WDT_2_18_CYCLES,
-	WDT_2_19_CYCLES,
-	WDT_2_20_CYCLES,
-	WDT_2_21_CYCLES,
-	WDT_2_22_CYCLES,
-	WDT_2_23_CYCLES,
-	WDT_2_24_CYCLES,
-	WDT_2_25_CYCLES,
-	WDT_2_26_CYCLES,
-	WDT_2_27_CYCLES,
-	WDT_2_28_CYCLES,
-	WDT_2_29_CYCLES,
-	WDT_2_30_CYCLES,
-	WDT_2_31_CYCLES
+struct wdt_cfg {
+	u8_t options;
 };
 
+/** Watchdog callback */
+typedef void (*wdt_callback_t)(struct device *dev, void *handle);
 
 /**
- * WDT configuration struct.
+ * @brief Watchdog timeout configuration struct.
+ *
+ * @param window - timing parameters of watchdog timeout.
+ * @param handle - unique handle used to reload specific watchdog timeout.
+ *		   This handle will also be supplied to callback if
+ *		   WDT_MODE_CALLBACK is set.
+ * @param callback - timeout callback.
+ * @param next - pointer to next timeout configuration. It is used for
+ *		 watchdogs with staged timeouts functionality. Should be NULL
+ *		 for single stage timeout.
  */
-struct wdt_config {
-	u32_t timeout;
-	enum wdt_mode mode;
-	void (*interrupt_fn)(struct device *dev);
+struct wdt_timeout_cfg {
```

Comment:
@mnkp I was also thinking about that. The question is if that API should be so robust or maybe we should left configuring the behavior of each stage in Kconfig.SoC. Suppose that WDT_MODE_CALLBACK is not set so there generally will be some kind of reset.
What kind of reset and after what stage could be configured in Kconfig. Then driver would only check if configuring timeouts is consistent with options set in Kconfig. For example if following option is desired:
    

- stage 1 -> run callback_1
-  stage 2 -> run callback_2
-  stage 3 -> perform SoC reset

the driver would return some error code if not enough `wdt_timeout_cfg` are supplied.

Reply:
There are a few problems with using Kconfig:
- It will make the API opaque. User will have difficulty determining what is happening.
- it will not be very portable. Either every driver will implement their own set of Kconfig options or we try to come up with a common set of options defined in drivers/watchdog/Kconfig. The former is not portable, in the latter case we won't be able to map the options between different driver instances. Unless we put it in DTS but it's not straightforward. I think this should be considered the last resort.